### DISCUSSION TITLE
How do I incorporate QuickLint?

### DISCUSSION THREAD

**Discussion by skywalkerdude** [Quick/Quick#1314] [Q&A] [answered]
Could someone write or point me to instructions on how to use [QuickLint](https://github.com/Quick/Quick/releases/tag/v7.6.0) with SPM?

Specifically:

1. There are two quicklint packages. Should I be importing both of them?
![Screenshot 2024-12-09 at 4 02 21 PM](https://github.com/user-attachments/assets/c5138a81-6d67-4d5e-a2f2-020a3ee3a9d0)
2. Is it possible to automate it with XCode by adding them as part of the "Run Build Tool Plug-ins" build phase?
![Screenshot 2024-12-09 at 4 03 56 PM](https://github.com/user-attachments/assets/2b830907-42ad-41e8-83c5-e47826d1d188)
3. If so, should I be using `LintError` or `LintWarning` as the package plug-in to use?
![Screenshot 2024-12-09 at 4 05 58 PM](https://github.com/user-attachments/assets/946dac34-3ed9-458d-9126-ea75c9a6794c)

**Comment by younata** [comment, marked as answer]
That documentation sadly doesn't exist elsewhere than... I guess right here. Sorry for not writing any beforehand.

quicklint (lowercase) is a cli executable. It's meant more for CI, though it can be used by developers. It has 2 subcommands: `lint` (the default) and `defocus`. The `lint` command will mark instances of focused examples as warnings (if you passing in the `--error` flag, it'll mark them as errors & return a non-zero exit code if any are found). The `defocus` command will attempt to fix remove focus from any focused examples. This is command is meant to be run by developers (or maybe incorporated as part of a pre-commit hook).

QuickLint (capitalized) is an alias to the lowercased one, and in a future version will likely be removed or turned into an internal library. Whenever I get find time & motivation to do so.

There are also the package plugin tools: LintWarning, LintError, LintCommand, and DefocusCommand.
LintWarning & LintError are the same as setting up an xcode run script to call `quicklint lint`, with the only difference being whether instances of focused commands being marked as warning (LintWarning) or errors (LintError). Both of these are implemented as build plugins, which means they can run automatically for you during the build phase (the only reason they aren't prebuild commands is because that would require me to ship binaries of QuickLint, which I haven't found the time & motivation to set up the infrastructure to do so).
LintCommand and DefocusCommand are xcode package commands, and allow you to invoke quicklint from xcode. LintCommand does what you expect, outputting warnings if any focused specs are found. DefocusCommand does the same thing as `quicklint defocus`.

I would recommend one of the following:

- Set up a run script command in xcode that runs `quicklint lint`. This run script command should detect if it's running in CI (which can generally be done by checking if the CI environment variable is defined, but you'll have to double check that yourself), and, if so, add in the `--error` flag. This way, developers can focus examples without getting build failures, but CI will correctly fail the build if a developer forgets to omit those.

OR

Set up LintWarning as an xcode build command so that developers can focus examples without having build errors.
Set up CI to run `quicklint error` in the test build phase so that developers don't accidentally merge in focused examples.

In either case, it may also be a good idea to add a pre-commit hook that runs `quicklint defocus` and adds the defocused files to the git. This is tricky to do well, because it might also include files/chunks you didn't mean to include (which is why I'm merely stating it's a good idea, not that it's a full recommendation for use. This is also why I recommend having CI error & fail the build if it detects focused examples).

There are also potential issues with running `quicklint` as a script, especially given Xcode's implementation of Swift Package Manager. I haven't tested this out, so I may be wrong, but one way to do that is to have a separate tools package that includes your SPM tools (such as Quick). You should be able to invoke quicklint by running `swift run quicklint`, with the working directory set to wherever that tools package is. You will also have to specify the directory of your project (likely the parent directory?). Again, I haven't tested this, so there may be other issues with doing this.
